storage/range_command.go

```diff
@@ -567,10 +567,11 @@ func (r *Range) InternalPushTxn(batch engine.Engine, ms *proto.MVCCStats, args *
 	if reply.PusheeTxn.LastHeartbeat == nil {
 		reply.PusheeTxn.LastHeartbeat = &reply.PusheeTxn.Timestamp
 	}
-	// Compute heartbeat expiration (use args.Timestamp instead of node
-	// clock wall time so that all replicas see the same result).
-	expiry := args.Timestamp
+	// Compute heartbeat expiration (all replicas must see the same result).
+	expiry := args.Now             // caller can set this to his wall time.
+	expiry.Forward(args.Timestamp) // if Now is not set, fallback
```

Comment:
Is there any reason to send a push txn request without setting `Now`? It might be simpler to require that `Now` be set in all cases. (On the other hand, this does ensure that we use `max(Now, Timestamp)` if there is any risk that the pusher's clock could be behind)